test/runtests.jl

```diff
@@ -1,6 +1,192 @@
+using AxisKeys
 using AxisSets
+using Dates
+using Impute
+using Missings
+using ReadOnlyArrays
+using Tables
 using Test
+using TimeZones
+
+using AxisSets: Dataset
 
 @testset "AxisSets.jl" begin
-    # Write your tests here.
+    @testset "Construction" begin
```

Comment:
these are great, I think it would be nice to support nested Tuples/NamedTuples along the likes of

```julia
data = (
    val1 = (a1=A1, a2=A2,... ), 
    val2 = (b1=B1, b2=B2,...),
   ...
)

Dataset(data, :time, :loc, :obj)  # using Tables constructor below I guess?
```
which would return a data set like
```julia
dataset.dims == OrderedSet(:time, :loc, :obj)
collect(keys(dataset.data)) == [:val1_a1, :val1_a2, ... :val2_b1, :val2_b2, ...] 
```

does that make sense? That should be doable with the Tables constructor I assume?

Reply:
Hmm, I'm not sure that should be handled by the constructor. I'm inclined to say that we want a kind of `flatten` function which just takes:

```julia
julia> data = (
    val1 = (a1=1, a2=2), 
    val2 = (b1=11, b2=22),
)
(val1 = (a1 = 1, a2 = 2), val2 = (b1 = 11, b2 = 22))

julia> flatten(data)
(val1_a1 = 1, val1_a2 = 2, val2_b1 = 11, val2_b2 = 22)
```

Then you could just use the existing constructors with `Dataset(dims...; flatten(data)...)`.

NOTE: `Iterators.flatten` returns an iterator over the flattened values, but in our case we want a `NamedTuple` with the correct merged keys back, so we'd need to implement it ourselves in a utility function for now.

Reply:
Alternatively, if we'd rather our keys be tuples then we could have it return:

```julia
((val1, a1) = 1, (val1, a2) = 2, (val2, b1) = 11, (val2, b2) = 22)
```

Reply:
Okay, regarding the nested structure approach, it looks like we should be able to support both the symbol flattening and the partial tuple indexing:

From kwargs needing symbol flattening:
```julia
julia> data = (
          group1 = (
              a = KeyedArray(
                  rand(4, 3, 2);
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  loc=1:3,
                  obj=[:a, :b],
              ),
              b = KeyedArray(
                  rand(4, 2);
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  label=["x", "y"],
              )
          ),
          group2 = (
              a = KeyedArray(
                  rand(4, 3, 2) .+ 1.0;
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  loc=1:3,
                  obj=[:a, :b],
              ),
              b = KeyedArray(
                  rand(4, 2) .+ 1.0;
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  label=["x", "y"],
              )
          )
      );

julia> ds = Dataset(; dims=(:time, :loc, :obj, :label), flatten(data)...)
Dataset{Symbol}(; dims=(:time, :loc, :obj, :label)) with 4 entries:
  group1_a => 3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   loc ∈ 3-element UnitRange{Int64}
    □   obj ∈ 2-element Vector{Symbol}
    And data, 4×3×2 Array{Float64,3}
    [:, :, 1] =
                                         (1)                    (2)                    (3)
       DateTime("2021-01-01T11:00:00")     0.04872510414699738    0.29110518330009927    0.2986134227434931
       DateTime("2021-01-01T12:00:00")     0.7762864652411638     0.35320877889972513    0.31848456844374584
       DateTime("2021-01-01T13:00:00")     0.46016742704989877    0.09777842534184167    0.03205479156317814
       DateTime("2021-01-01T14:00:00")     0.7156851812256597     0.3469152902099679     0.7940667604433369

    [:, :, 2] =
                                         (1)                    (2)                    (3)
       DateTime("2021-01-01T11:00:00")     0.2874908306388013     0.11561259526797163    0.04106664371226665
       DateTime("2021-01-01T12:00:00")     0.11290997398628821    0.5198660452284298     0.4174890858753668
       DateTime("2021-01-01T13:00:00")     0.03395272449147235    0.17277356472554528    0.5449761382674334
       DateTime("2021-01-01T14:00:00")     0.657367396272627      0.9659217386734        0.9488544077075811
  group1_b => 2-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   label ∈ 2-element Vector{String}
    And data, 4×2 Array{Float64,2}
                                         ("x")                ("y")
       DateTime("2021-01-01T11:00:00")    0.3762703160874752   0.6504372664055247
       DateTime("2021-01-01T12:00:00")    0.5996244048268149   0.3703834055867996
       DateTime("2021-01-01T13:00:00")    0.7931542362994977   0.5199667144017202
       DateTime("2021-01-01T14:00:00")    0.9900544652542147   0.11013346289002324
  group2_a => 3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   loc ∈ 3-element UnitRange{Int64}
    □   obj ∈ 2-element Vector{Symbol}
    And data, 4×3×2 Array{Float64,3}
    [:, :, 1] =
                                         (1)                   (2)                   (3)
       DateTime("2021-01-01T11:00:00")     1.3612977210403485    1.453087252869303     1.4572127985540482
       DateTime("2021-01-01T12:00:00")     1.340821761660883     1.4118399719350974    1.1853265030942535
       DateTime("2021-01-01T13:00:00")     1.5206079890816127    1.1403801567434522    1.983121705390562
       DateTime("2021-01-01T14:00:00")     1.6086968196136804    1.57958260726942      1.1860338042422265

    [:, :, 2] =
                                         (1)                   (2)                   (3)
       DateTime("2021-01-01T11:00:00")     1.2583539956404943    1.2812789562140843    1.9298671484667218
       DateTime("2021-01-01T12:00:00")     1.719975811423362     1.9733271829559769    1.4235615888880921
       DateTime("2021-01-01T13:00:00")     1.3764869400654414    1.530992321617456     1.2651468113069393
       DateTime("2021-01-01T14:00:00")     1.3714488060667684    1.2895141236503203    1.2406107888953029
  group2_b => 2-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   label ∈ 2-element Vector{String}
    And data, 4×2 Array{Float64,2}
                                         ("x")                ("y")
       DateTime("2021-01-01T11:00:00")    1.6207928184445002   1.2718183710804716
       DateTime("2021-01-01T12:00:00")    1.0952601311481949   1.010272360030641
       DateTime("2021-01-01T13:00:00")    1.4767716208219948   1.1551741048917705
       DateTime("2021-01-01T14:00:00")    1.3696418775758623   1.1474687325874362

julia> ds[:group2_b]
2-dimensional KeyedArray(NamedDimsArray(...)) with keys:
↓   time ∈ 4-element ReadOnlyArray{DateTime,...}
→   label ∈ 2-element ReadOnlyArray{String,...}
And data, 4×2 Array{Float64,2}:
                                     ("x")     ("y")
   DateTime("2021-01-01T11:00:00")    1.62079   1.27182
   DateTime("2021-01-01T12:00:00")    1.09526   1.01027
   DateTime("2021-01-01T13:00:00")    1.47677   1.15517
   DateTime("2021-01-01T14:00:00")    1.36964   1.14747
```

Tuple keys and partial indexing:
```julia
julia> data = [
          :group1 => [
              :a => KeyedArray(
                  rand(4, 3, 2);
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  loc=1:3,
                  obj=[:a, :b],
              ),
              :b => KeyedArray(
                  rand(4, 2);
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  label=["x", "y"],
              )
          ],
          :group2 => [
              :a => KeyedArray(
                  rand(4, 3, 2) .+ 1.0;
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  loc=1:3,
                  obj=[:a, :b],
              ),
              :b => KeyedArray(
                  rand(4, 2) .+ 1.0;
                  time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
                  label=["x", "y"],
              )
          ]
      ];

julia> ds = Dataset(flatten(data)...; dims=(:time, :loc, :obj, :label))
Dataset{Tuple{Symbol,Symbol}}(; dims=(:time, :loc, :obj, :label)) with 4 entries:
  (:group1, :a) => 3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   loc ∈ 3-element UnitRange{Int64}
    □   obj ∈ 2-element Vector{Symbol}
    And data, 4×3×2 Array{Float64,3}
    [:, :, 1] =
                                         (1)                    (2)                   (3)
       DateTime("2021-01-01T11:00:00")     0.08806467933919859    0.5345395498067949    0.21600006066199184
       DateTime("2021-01-01T12:00:00")     0.8741648478256292     0.7394674963505965    0.22924683051312544
       DateTime("2021-01-01T13:00:00")     0.5839519011175041     0.9582076982048542    0.4637947371817779
       DateTime("2021-01-01T14:00:00")     0.43963180603848895    0.9593683281550414    0.04081125661907037

    [:, :, 2] =
                                         (1)                   (2)                    (3)
       DateTime("2021-01-01T11:00:00")     0.5975954957465563    0.4247189218250438     0.25928482797858576
       DateTime("2021-01-01T12:00:00")     0.901654953988529     0.567846976844183      0.2716367784163576
       DateTime("2021-01-01T13:00:00")     0.2665648605084192    0.36261637456029483    0.6985572269809432
       DateTime("2021-01-01T14:00:00")     0.8275741335779658    0.5941971585629624     0.20254316041710885
  (:group1, :b) => 2-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   label ∈ 2-element Vector{String}
    And data, 4×2 Array{Float64,2}
                                         ("x")                 ("y")
       DateTime("2021-01-01T11:00:00")    0.6041738562945647    0.9124002300037084
       DateTime("2021-01-01T12:00:00")    0.4388707350583372    0.8065757450855833
       DateTime("2021-01-01T13:00:00")    0.22817380792382935   0.7297136871413197
       DateTime("2021-01-01T14:00:00")    0.6254928621493481    0.47824971845436215
  (:group2, :a) => 3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   loc ∈ 3-element UnitRange{Int64}
    □   obj ∈ 2-element Vector{Symbol}
    And data, 4×3×2 Array{Float64,3}
    [:, :, 1] =
                                         (1)                   (2)                   (3)
       DateTime("2021-01-01T11:00:00")     1.718583715679297     1.6139324639697805    1.433521830143539
       DateTime("2021-01-01T12:00:00")     1.3119552067960423    1.4751347688932521    1.7308039123391237
       DateTime("2021-01-01T13:00:00")     1.9689987861918545    1.5961827555013055    1.5949909132671534
       DateTime("2021-01-01T14:00:00")     1.386373865828174     1.1841865266931793    1.9700928776538524

    [:, :, 2] =
                                         (1)                   (2)                   (3)
       DateTime("2021-01-01T11:00:00")     1.2727801736140825    1.7206581204625575    1.4372125484499394
       DateTime("2021-01-01T12:00:00")     1.853774110152573     1.9253785723869947    1.3226005088661816
       DateTime("2021-01-01T13:00:00")     1.880828311670044     1.5075657985556132    1.5833382865645342
       DateTime("2021-01-01T14:00:00")     1.154320511516146     1.4316812144070397    1.592149589885552
  (:group2, :b) => 2-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element StepRange{DateTime,...}
    →   label ∈ 2-element Vector{String}
    And data, 4×2 Array{Float64,2}
                                         ("x")                ("y")
       DateTime("2021-01-01T11:00:00")    1.6478180180668485   1.4865984413720192
       DateTime("2021-01-01T12:00:00")    1.3259764858753564   1.3932010013767298
       DateTime("2021-01-01T13:00:00")    1.8147090379827169   1.7100990164373135
       DateTime("2021-01-01T14:00:00")    1.3485814045041327   1.2696486405639187

julia> ds[:a]
Dataset{Tuple{Symbol,Symbol}}(; dims=(:time, :loc, :obj)) with 2 entries:
  (:group1, :a) => 3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element ReadOnlyArray{DateTime,...}
    →   loc ∈ 3-element ReadOnlyArray{Int64,...}
    □   obj ∈ 2-element ReadOnlyArray{Symbol,...}
    And data, 4×3×2 Array{Float64,3}
    [:, :, 1] =
                                         (1)                    (2)                   (3)
       DateTime("2021-01-01T11:00:00")     0.08806467933919859    0.5345395498067949    0.21600006066199184
       DateTime("2021-01-01T12:00:00")     0.8741648478256292     0.7394674963505965    0.22924683051312544
       DateTime("2021-01-01T13:00:00")     0.5839519011175041     0.9582076982048542    0.4637947371817779
       DateTime("2021-01-01T14:00:00")     0.43963180603848895    0.9593683281550414    0.04081125661907037

    [:, :, 2] =
                                         (1)                   (2)                    (3)
       DateTime("2021-01-01T11:00:00")     0.5975954957465563    0.4247189218250438     0.25928482797858576
       DateTime("2021-01-01T12:00:00")     0.901654953988529     0.567846976844183      0.2716367784163576
       DateTime("2021-01-01T13:00:00")     0.2665648605084192    0.36261637456029483    0.6985572269809432
       DateTime("2021-01-01T14:00:00")     0.8275741335779658    0.5941971585629624     0.20254316041710885
  (:group2, :a) => 3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element ReadOnlyArray{DateTime,...}
    →   loc ∈ 3-element ReadOnlyArray{Int64,...}
    □   obj ∈ 2-element ReadOnlyArray{Symbol,...}
    And data, 4×3×2 Array{Float64,3}
    [:, :, 1] =
                                         (1)                   (2)                   (3)
       DateTime("2021-01-01T11:00:00")     1.718583715679297     1.6139324639697805    1.433521830143539
       DateTime("2021-01-01T12:00:00")     1.3119552067960423    1.4751347688932521    1.7308039123391237
       DateTime("2021-01-01T13:00:00")     1.9689987861918545    1.5961827555013055    1.5949909132671534
       DateTime("2021-01-01T14:00:00")     1.386373865828174     1.1841865266931793    1.9700928776538524

    [:, :, 2] =
                                         (1)                   (2)                   (3)
       DateTime("2021-01-01T11:00:00")     1.2727801736140825    1.7206581204625575    1.4372125484499394
       DateTime("2021-01-01T12:00:00")     1.853774110152573     1.9253785723869947    1.3226005088661816
       DateTime("2021-01-01T13:00:00")     1.880828311670044     1.5075657985556132    1.5833382865645342
       DateTime("2021-01-01T14:00:00")     1.154320511516146     1.4316812144070397    1.592149589885552

julia> ds[:group1]
Dataset{Tuple{Symbol,Symbol}}(; dims=(:time,)) with 2 entries:
  (:group1, :a) => 3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element ReadOnlyArray{DateTime,...}
    →   loc ∈ 3-element ReadOnlyArray{Int64,...}
    □   obj ∈ 2-element ReadOnlyArray{Symbol,...}
    And data, 4×3×2 Array{Float64,3}
    [:, :, 1] =
                                         (1)                    (2)                   (3)
       DateTime("2021-01-01T11:00:00")     0.08806467933919859    0.5345395498067949    0.21600006066199184
       DateTime("2021-01-01T12:00:00")     0.8741648478256292     0.7394674963505965    0.22924683051312544
       DateTime("2021-01-01T13:00:00")     0.5839519011175041     0.9582076982048542    0.4637947371817779
       DateTime("2021-01-01T14:00:00")     0.43963180603848895    0.9593683281550414    0.04081125661907037

    [:, :, 2] =
                                         (1)                   (2)                    (3)
       DateTime("2021-01-01T11:00:00")     0.5975954957465563    0.4247189218250438     0.25928482797858576
       DateTime("2021-01-01T12:00:00")     0.901654953988529     0.567846976844183      0.2716367784163576
       DateTime("2021-01-01T13:00:00")     0.2665648605084192    0.36261637456029483    0.6985572269809432
       DateTime("2021-01-01T14:00:00")     0.8275741335779658    0.5941971585629624     0.20254316041710885
  (:group1, :b) => 2-dimensional KeyedArray(NamedDimsArray(...)) with keys:
    ↓   time ∈ 4-element ReadOnlyArray{DateTime,...}
    →   label ∈ 2-element ReadOnlyArray{String,...}
    And data, 4×2 Array{Float64,2}
                                         ("x")                 ("y")
       DateTime("2021-01-01T11:00:00")    0.6041738562945647    0.9124002300037084
       DateTime("2021-01-01T12:00:00")    0.4388707350583372    0.8065757450855833
       DateTime("2021-01-01T13:00:00")    0.22817380792382935   0.7297136871413197
       DateTime("2021-01-01T14:00:00")    0.6254928621493481    0.47824971845436215
```

Reply:
awesome! 😎

Reply:
NOTE: I've recently deciding to replace the default `:_` separator with `:ᵡ` because we've had some conflicts where `_` often isn't unique when search for a flattened group. Also, we can now flatten dimensions for `KeyedArrays` in a similar way. We should also be able to support `flatten` operations across components of a `Dataset`.
```julia
julia> A = KeyedArray(
                       reshape(1:24, (4, 3, 2));
                       time=dt,
                       loc=1:3,
                       obj=[:a, :b],
                   )
3-dimensional KeyedArray(NamedDimsArray(...)) with keys:
↓   time ∈ 4-element StepRange{DateTime,...}
→   loc ∈ 3-element UnitRange{Int64}
□   obj ∈ 2-element Vector{Symbol}
And data, 4×3×2 reshape(::UnitRange{Int64}, 4, 3, 2) with eltype Int64:
[:, :, 1] ~ (:, :, :a):
                                     (1)  (2)  (3)
   DateTime("2021-01-01T11:00:00")     1    5    9
   DateTime("2021-01-01T12:00:00")     2    6   10
   DateTime("2021-01-01T13:00:00")     3    7   11
   DateTime("2021-01-01T14:00:00")     4    8   12

[:, :, 2] ~ (:, :, :b):
                                     (1)  (2)  (3)
   DateTime("2021-01-01T11:00:00")    13   17   21
   DateTime("2021-01-01T12:00:00")    14   18   22
   DateTime("2021-01-01T13:00:00")    15   19   23
   
julia> flatten(A, (:loc, :obj))
2-dimensional KeyedArray(NamedDimsArray(...)) with keys:
↓   time ∈ 4-element StepRange{DateTime,...}
→   locᵡobj ∈ 6-element Vector{Symbol}
And data, 4×6 reshape(::UnitRange{Int64}, 4, 6) with eltype Int64:
                                      Symbol("1ᵡa")    Symbol("2ᵡa")    Symbol("3ᵡa")    Symbol("1ᵡb")    Symbol("2ᵡb")    Symbol("3ᵡb")
   DateTime("2021-01-01T11:00:00")    1                5                9               13               17               21
   DateTime("2021-01-01T12:00:00")    2                6               10               14               18               22
   DateTime("2021-01-01T13:00:00")    3                7               11               15               19               23
   DateTime("2021-01-01T14:00:00")    4                8               12               16               20               24
```